packages/react/src/components/header/components/headerActionBar/HeaderActionBar.tsx

```diff
@@ -9,9 +9,70 @@ import { useCallbackIfDefined, useEnterOrSpacePressCallback } from '../../../../
 import { HeaderActionBarMenuItem } from '../headerActionBarItem';
 import styles from './HeaderActionBar.module.scss';
 import HeaderActionBarLogo from './HeaderActionBarLogo';
+import { useHeaderContext } from '../../HeaderContext';
 
 const classNames = styleBoundClassNames(styles);
 
+enum TabBarrierPosition {
+  top = 'top',
+  bottom = 'bottom',
+}
+
+type TabBarrierProps = {
+  id: string;
+  tabIndex: number;
+  'aria-hidden': boolean;
+};
+
+const defaultBarrierProps: Partial<TabBarrierProps> = {
+  tabIndex: 0,
+  'aria-hidden': true,
+};
+
+const elementIsInViewport = (element: Element) => {
+  const { top, left, bottom, right } = element.getBoundingClientRect();
+  const { innerHeight, innerWidth } = window;
+
+  return top > 0 && left > 0 && bottom < innerHeight && right < innerWidth;
+};
+
+const findFocusableElementsWithin = (element: HTMLElement) => {
+  const focusable = element.querySelectorAll('a, button, textarea, input[type="text"], select');
+
+  return Array.from(focusable).filter((item) => elementIsInViewport(item));
+};
```

Comment:
This should also check `tablndex`-attribute to be 100% correct. And is element disabled. And is it hidden (width and height > 0). The `elementIsInViewport` will return visually hidden elements too.

Devs can add anything to the menus, so plain element type check is not accurate enough imho.


I'd also make this a common util function, but then the elementIsInViewport(item) should be discarded, because being outside of viewport does not make an element unfocusable. It is of course correct in this context inside the menu.

Reply:
Good thinking 👍 https://github.com/City-of-Helsinki/helsinki-design-system/pull/1120/commits/6468f7395b11b4f238fa248af14a93bd3fe5fcbc